genetic_algorithms/algorithms/hillclimbing.py

```diff
@@ -1,16 +1,44 @@
 from genetic_algorithms.models.algorithm import Algorithm
 from genetic_algorithms.problems.base.state import State
 from genetic_algorithms.problems.base.model import Model
+from genetic_algorithms.helpers import History
+from typing import Tuple
+
+HISTORY_SIZE = 5
 
 
 class HillClimbing(Algorithm):
-    def next_state(self, model: Model, state: State) -> State:
+    def __init__(self):
+        super(HillClimbing, self).__init__()
+        self.history = History[int](HISTORY_SIZE)
+        self.n_restarts = 0
+
+    def _hill_climbing(self, model: Model, state: State) -> Tuple[State, int]:
         best_state = state
         best_state_cost = model.cost_for(state)
         for move in model.move_generator.available_moves(state):
             new_state = move.make()
             new_state_cost = model.cost_for(new_state)
             if new_state_cost <= best_state_cost:
                 (best_state, best_state_cost) = (new_state, new_state_cost)
-        return best_state
+        return best_state, best_state_cost
+
+    def next_state(self, model: Model, state: State) -> State:
+        next_state, next_state_cost = self._hill_climbing(model, state)
+        self.history.append(next_state_cost)
+        if self._is_extremum(next_state_cost):
+            if self.n_restarts > 0:
+                next_state = self._restart(model)
+                self.n_restarts -= 1
+            else:
```

Comment:
To mi się trochę nie podoba, ponieważ domyślnie w `HillClimbing` nie ma czegoś takiego jak `restarty` (co zresztą widać z pustej metody `_restart`), a ten kawałek kodu odpowiada raczej `IteratedLocalSearch` - robisz restart kiedy napotkasz ekstremum.



Zamiast robić tak, ja bym nadpisał metodę `next_state` w każdym z dziedziczących algorytmów.
Ewentualnie, żeby się nie powtarazać, można dodać `hook`: `_on_extremum` i wywoływać go tutaj, która domyślnie będzie ustawiać `self._is_terminated` na `True`, a potem w każej z implementacji, można będzie go nadpisać.